Add GetPreferredAllocation() call to the v1beta1 device plugin API

klueska opened the pull request:
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test

/kind feature

> /kind flake

**What this PR does / why we need it**:
This PR adds support for a new, optional call in the `v1beta1` device plugin API called `GetPreferredAllocation()`. Because it is an optional extension (and not a change) to the API, it does not require a version bump.

Details of this new API call can be found in the following KEP update: https://github.com/kubernetes/enhancements/pull/1121

Please review commit-by-commit.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #83483

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Added a new `GetPreferredAllocation()` call to the `v1beta1` device plugin API.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
[KEP]: https://github.com/kubernetes/enhancements/pull/1121
```

klueska commented:
/assign @derekwaynecarr 
To approve the API changes

The only reason this this currently a WIP is because I want to add more extensive testing before merging. We should be able to work on approving the API changes without this though.

klueska commented:
/assign @mkumatag 
/assign @listx
For approval in small change to `test/images/OWNERS` in support of this new API.

klueska commented:
/retest

klueska commented:
/test pull-kubernetes-e2e-kind

klueska commented:
/retest

nolancon commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -675,32 +674,69 @@ func (m *ManagerImpl) devicesToAllocate(podUID, contName, resource string, requi
+	// Filters available Devices based on NUMA affinity.
+	aligned, unaligned, noAffinity := m.filterByAffinity(podUID, contName, resource, available)
+
+	// Allocates from aligned Devices list.
+	for device := range aligned.Difference(devices) {
+		m.allocatedDevices[resource].Insert(device)
+		devices.Insert(device)
+		needed--
+		if needed == 0 {
+			return devices, nil
+		}
+	}
+	// Calls GetPreferredAllocation if it is available.
+	preferred, err := m.callGetPreferredAllocationIfAvailable(podUID, contName, resource, available.Union(reusableDevices), devices, required)
+	if err != nil {
+		return nil, err
+	}
+	// Allocates from the preferred Devices list.
+	for device := range preferred.Difference(devices) {
+		m.allocatedDevices[resource].Insert(device)
+		devices.Insert(device)
+		needed--
+		if needed == 0 {
+			return devices, nil
+		}
+	}
```
I would have thought allocation from the preferred list would take place before the aligned list. Is this because we can't assume preferred devices are aligned also?

klueska replied:
Exactly. The goal is to first honour allocations based on topologymanager alignment, and only then defer to the plugin to help decide how to alllocate the rest.

klueska replied:
Updated with detailed comments in latest

nolancon commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -920,6 +956,33 @@ func (m *ManagerImpl) callPreStartContainerIfNeeded(podUID, contName, resource s
+		return nil, nil
+	}
+
+	devices := m.podDevices.containerDevices(podUID, contName, resource)
+	if devices == nil {
+		return nil, fmt.Errorf("no devices found allocated in local cache for pod %s, container %s, resource %s", podUID, contName, resource)
+	}
```
Could you explain why this check and returning an error is necessary here? It's not immediately clear to me, thanks.

klueska replied:
I didn't actually look too closely at this when adding it.

It is copy-pasted from the `callPreStartContainerIfNeeded()` function above, which is similar in nature to this one (i.e. only call out to the plugin it if the options set on the plugin indicate to).

Let me take a closer look at everything in this function to see if we need it.

klueska replied:
Just looked. Good catch. Not only is it unnecessary, it would have caused problems once I started writing tests that actually exercise this function (because it's highly likely that no devices will be allocated at the time this is called!).

klueska replied:
Fixed in latest

nolancon reviewed (commented):
Just one nit and a couple of questions for my own understanding.

vpickard reviewed (commented):
Looks good, few minor comment. Thanks!

klueska commented on pkg/kubelet/cm/devicemanager/device_plugin_stub.go:
```diff
@@ -47,11 +48,23 @@ type Stub struct {
 
 }
 
+// stubGetPreferredAllocationFunc is the function called when receive an getPreferredAllocation request from Kubelet
```
Done

klueska commented on pkg/kubelet/cm/devicemanager/device_plugin_stub.go:
```diff
@@ -47,11 +48,23 @@ type Stub struct {
 
 }
 
+// stubGetPreferredAllocationFunc is the function called when receive an getPreferredAllocation request from Kubelet
```
Done

nolancon reviewed (commented):
One question. Changes look good otherwise.

klueska commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -675,32 +674,103 @@ func (m *ManagerImpl) devicesToAllocate(podUID, contName, resource string, requi
+
+	// If we can allocate all remaining devices from the set of aligned ones, then
+	// give the plugin the chance to influence which ones to allocate from that set.
+	if needed < aligned.Len() {
```
No. If they are equal then I want to fall through and just allocate all of the needed ones from the aligned set directly.

klueska committed:
Update whitespace in device plugin api.proto

klueska committed:
Fix some whitespacing and comments in devicemanager

klueska committed:
Add GetPreferredAllocation() call to the device plugin api.proto

The details of this API can be found in:
https://github.com/kubernetes/enhancements/pull/1121

klueska committed:
Update device plugin stub with new GetPreferredAllocation() call

klueska committed:
Add GetPreferredAllocation() as a supported device plugin endpoint

klueska committed:
Add callGetPreferredAllocationIfAvailable() function in devicemanager

This function mimics what is already done for the conditional call to
PreStartContainer() via the callPreStartContainerIfNeeded() function.

klueska committed:
Updates logic in devicesToAllocate() to call GetPreferredAllocation()

klueska committed:
Update existing tests in support of GetPreferredallocation()

klueska committed:
Add a simple endpoint test for GetPreferredAllocation()

More extensive tests that exercise the allocation logic are to follow.

klueska committed:
Remove unnecessary field from TestTopologyAlignedAllocation() test cases

klueska committed:
Add new test cases for GetPreferredAllocation() in allocation path

klueska commented:
/retest

derekwaynecarr commented:
nice job, tests look good too.

/approve
/lgtm

klueska commented:
/retest

klueska commented:
/retest

klueska commented:
/retest

mkumatag reviewed (changes requested):
Please bump the version here: https://github.com/kubernetes/kubernetes/blob/master/test/images/sample-device-plugin/VERSION

klueska commented:
> Please bump the version here: https://github.com/kubernetes/kubernetes/blob/master/test/images/sample-device-plugin/VERSION

Thanks for the review. VERSION has been updated here: https://github.com/kubernetes/kubernetes/pull/92665/commits/cbd405d85cca84344f3acc21c86a9eff00ff58b2#diff-efcabe360f36f7a76db00afca0404498

klueska commented:
/retest

klueska commented:
/retest

mkumatag commented:
/lgtm
/approve
/hold

Only for `k/k/test/images` content, adding hold for the remaining

klueska commented:
/retest

klueska commented:
/hold cancel
